File: languages/python/reverse-complement.py
import sys

P = 1000000007
IM = 139968
IA = 3877
IC = 29573


def comp(c):  # A<->T, C<->G; only A/C/G/T occur
    if c == 65:  # 'A'
        return 84  # 'T'
    if c == 67:  # 'C'
        return 71  # 'G'
    if c == 71:  # 'G'
        return 67  # 'C'
    return 65  # 'A'


def reverse_complement(L):
    s = bytearray(L)
    seed = 42
    for i in range(L):
        seed = (seed * IA + IC) % IM
        if seed < 42000:
            s[i] = 65  # 'A'
        elif seed < 70000:
            s[i] = 67  # 'C'
        elif seed < 98000:
            s[i] = 71  # 'G'
        else:
            s[i] = 84  # 'T'

    i = 0
    j = L - 1
    while i < j:  # two-pointer reverse-and-complement, in place
        a = comp(s[i])
        s[i] = comp(s[j])
        s[j] = a
        i += 1
        j -= 1
    if i == j:
        s[i] = comp(s[i])  # middle char when L is odd

    h = 0
    for k in range(L):
        h = (h * 31 + s[k]) % P
    return h


if __name__ == "__main__":
    L = int(sys.argv[1]) if len(sys.argv) > 1 else 100000
    print(reverse_complement(L))
    print("reverse-complement(%d)" % L)

File: languages/javascript/reverse-complement.js
"use strict";

const P = 1000000007;
const IM = 139968;
const IA = 3877;
const IC = 29573;

function comp(c) {
  // A<->T, C<->G; only A/C/G/T occur
  if (c === 65) return 84; // 'A' -> 'T'
  if (c === 67) return 71; // 'C' -> 'G'
  if (c === 71) return 67; // 'G' -> 'C'
  return 65; // 'T' -> 'A'
}

function reverseComplement(L) {
  const s = new Uint8Array(L);
  let seed = 42;
  for (let i = 0; i < L; i++) {
    seed = (seed * IA + IC) % IM;
    if (seed < 42000) s[i] = 65; // 'A'
    else if (seed < 70000) s[i] = 67; // 'C'
    else if (seed < 98000) s[i] = 71; // 'G'
    else s[i] = 84; // 'T'
  }

  let i = 0;
  let j = L - 1;
  while (i < j) {
    // two-pointer reverse-and-complement, in place
    const a = comp(s[i]);
    s[i] = comp(s[j]);
    s[j] = a;
    i += 1;
    j -= 1;
  }
  if (i === j) s[i] = comp(s[i]); // middle char when L is odd

  let h = 0;
  for (let k = 0; k < L; k++) {
    h = (h * 31 + s[k]) % P;
  }
  return h;
}

function main() {
  const L = process.argv[2] !== undefined ? parseInt(process.argv[2], 10) : 100000;
  console.log(reverseComplement(L));
  console.log(`reverse-complement(${L})`);
}

main();
